### ISSUE TITLE
Loss computation question.

### ISSUE THREAD

**Issue by GYDDHPY** [mathpluscode/ImgX-DiffSeg#9]
Hi, thank you for your great job and sharing your code.
I have a question about the loss computation.
https://github.com/mathpluscode/ImgX-DiffSeg/blob/f3b2c3102493c2efb94556b08ae92e63179206d1/imgx/exp/loss.py#L237-L250

I notice that the `x_start` computated using the predicted noise is translated to logits using `x_to_logits` function. 
https://github.com/mathpluscode/ImgX-DiffSeg/blob/f3b2c3102493c2efb94556b08ae92e63179206d1/imgx/diffusion/gaussian_diffusion.py#L738-L753
 I think `x_start` can be regard as logits between -1 and 1, as the input `x_start` is in this format.  I cannot figure out why it is translated to logits again. Maybe I miss some point or misunderstand something. Could you please help me understand this operation?

Looking forward to your reply.

**Comment by mathpluscode** [comment]
Hi @GYDDHPY, thanks for your question. Please allow me to detail a bit about the thinking behind and hope the answers below can address your questions. Otherwise, please let us know if anything is not clear.

### Diffusion Model

First, let's define the diffusion model as below
<img width="641" alt="image" src="https://github.com/mathpluscode/ImgX-DiffSeg/assets/4066929/56638cc4-bc07-4214-a113-c1c6e7ae1043">
from which we can derive
<img width="236" alt="image" src="https://github.com/mathpluscode/ImgX-DiffSeg/assets/4066929/e11789c2-867e-423c-bc7a-723490a48bb8">
with
<img width="111" alt="image" src="https://github.com/mathpluscode/ImgX-DiffSeg/assets/4066929/3733db99-feb4-4fe9-a148-a7f1aab2ba18">

We use a linear beta schedule from 0.0001 to 0.02, therefore the `\bar{alpha}_t` will be very small, approaching zero for time T (fully noised). Thus we consider `x` approximately follows a standard normal distribution, most importantly, the space of `x` is symmetric and centered at zero.

Empirically, we found that the value of logits does not satisfy this symmetric assumption.

### Space of x

We consider `x` to have its own space, which can be
- scaled probabilities (2*probability-1), thus have value range [-1, 1]
- logits, thus have unbound range

**But we force the loss to be calculated with logits, therefore when `x` is scaled probability, we need to scale it back to [0, 1] and log it. Thus we need to call `x_to_logits`.**

In the code, the space of `x` is controlled by `x_space`. If you set it to `logits`, then `x_to_logits` or `logits_to_x` will be just dummy mapping.

Empirically, we found the scaled probability setting provides better results. This is reasonable, if we let `x` be logits, then `x` is from ground truth as during training and always have range in [-1, 1]. However, during inference, `x` is from model outputs, which will not be in [-1, 1] by default without specific control.

However, alternatively, we can clip the `x` to [-1, 1] in training. We haven't extensively studied this.


### Related work

This scaled probability setting is similar to the analog bits in https://arxiv.org/abs/2210.06366, where they further scaled `x` to be in [-b, b] with b <= 1.

### Model prediction

One side note, you mentioned "x_start computated using the predicted noise". In this work, we aim to send the message to the community that, for diffusion-based segmentation, it is better to task the model to predict the segmentation masks (`x_start`) instead of noise (`eps`) in mask spaces, therefore the model is not necessarily predicting noise.

However, as you may know, knowing the linear relation between (`eps`, `x_start`, `x_t`), knowing any two can derive the third missing one. So it's just different configurations.
<img width="249" alt="image" src="https://github.com/mathpluscode/ImgX-DiffSeg/assets/4066929/cfe5e1a7-c2aa-48ca-b8f8-a4cfcda8fa2f">

**Comment by GYDDHPY** [comment]
Thank you for your rapid reply and so detailed explanation.  
It seems I have a misunderstanding on `x_start` or `pred_start`. According to your explanation, `x_start` is also in probability format but just rescaled to a different range. I should not regard it as logits. Thus in loss computation, it should be transformed to logits again.
### Question 1
https://github.com/mathpluscode/ImgX-DiffSeg/blob/f3b2c3102493c2efb94556b08ae92e63179206d1/imgx/diffusion/gaussian_diffusion.py#L749-L753
I notice that in the `x_to_logits` function, you first rescale the value and then clip it and use `log` operation to get logits. I can understand the rescale operation, as the value input to the model is scaled to (-1,1) and the rescale operation here is a reverse process. However, the clip operation may ignore values out of the specified range `[EPS,1]`. Will this affect the computation results in your experiment?

### Question 2
The other question is about the `clip_x` operation in sampling.
https://github.com/mathpluscode/ImgX-DiffSeg/blob/f3b2c3102493c2efb94556b08ae92e63179206d1/imgx/diffusion/gaussian_diffusion.py#L344-L368

The `self.clip_x` is similar to operations in Improved DDPM:
https://github.com/openai/improved-diffusion/blob/783b6740edb79fdb7d063250db2c51cc9545dcd1/improved_diffusion/gaussian_diffusion.py#L293-L298
I notice that according to your settings, you did not clip the pre_start. This `clip` operation has been used in many diffusion related works and most of them use it in default, clip value to [-1,1].  However, I'm not sure whether this operation should be used in predicting segmentation mask tasks. I think the clip operation may ignore logits out of the specified range and lead to bad segmentation results. Why you choose not to use clip operation in your experiments? How do we decide the value range in clip operation?

Sorry for asking some questions about predicting noise, as I know that your work mainly focus on directly predicting segmentation mask.  

Looking forward to your reply.

**Comment by mathpluscode** [comment]
@GYDDHPY Hi, sorry for the late reply.

> However, the clip operation may ignore values out of the specified range [EPS,1]. Will this affect the computation results in your experiment?

The `x` is supposed to be in [-1, 1] therefore before clipping, probs should be in `[0, 1]`, there should be no values larger than 1. The clip to `EPS` is necessary as log of zero is not defined.

> Why you choose not to use clip operation in your experiments? How do we decide the value range in clip operation?

You are right, self.clip_x is the same as operations in Improved DDPM:
https://github.com/openai/improved-diffusion/blob/783b6740edb79fdb7d063250db2c51cc9545dcd1/improved_diffusion/gaussian_diffusion.py#L293-L298 or https://github.com/hojonathanho/diffusion/blob/1e0dceb3b3495bbe19116a5e1b3596cd0706c543/diffusion_tf/diffusion_utils_2.py#L151

My assumption is that in image generations, previous work mainly tasks the model to predict noise, as they often use 1000 steps of noising steps, the noise values should be small, therefore a clip is needed. In our case, as x is already bounded, I did not put the clip. The impact is not very clear, but it can be easily tested by changing the `x_limit`.

https://github.com/mathpluscode/ImgX-DiffSeg/blob/f3b2c3102493c2efb94556b08ae92e63179206d1/imgx/conf/config_amos_diffusion.yaml#L28

Hope this answers your questions :)

**Comment by mathpluscode** [comment]
Hi, I am closing this issue. Please feel free to reopen it if there are more questions. Thanks.
